library/src/commonMain/kotlin/com/badoo/reaktive/observable/Zip.kt

```diff
@@ -0,0 +1,231 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.disposable.CompositeDisposable
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.utils.lock.newLock
+import com.badoo.reaktive.utils.lock.synchronized
+
+fun <T, R> Collection<Observable<T>>.zip(mapper: (List<T>) -> R): Observable<R> =
+    observableByEmitter { emitter ->
+        val disposables = CompositeDisposable()
+        emitter.setDisposable(disposables)
+        val lock = newLock()
+        val values = List<MutableList<T>>(size) { mutableListOf() }
+        val readyValues = mutableListOf<T>()
+
+        forEachIndexed { index, source ->
+            source.subscribeSafe(
+                object : ObservableObserver<T> {
+                    override fun onSubscribe(disposable: Disposable) {
+                        disposables += disposable
+                    }
+
+                    override fun onNext(value: T) {
+                        lock.synchronized {
+                            values[index].add(value)
+
+                            if (values.all(List<*>::isNotEmpty)) {
+                                values.forEach { queue ->
+                                    readyValues.add(queue.removeAt(0))
+                                }
+
+                                mapper(readyValues)
+                                    .also {
+                                        readyValues.clear()
```

Comment:
Consider a mapper that simply returns the combined list ( (list) -> R = list )
When we call clear we would be clearing the output leading to unexpected behaviour. Better to copy values to a new list before emitting.

Reply:
I did not get this, could you explain?

Reply:
I've got it